src/blockchain_db/blockchain_db.cpp

```diff
@@ -77,35 +78,47 @@ bool matches_category(relay_method method, relay_category category) noexcept
 
 void txpool_tx_meta_t::set_relay_method(relay_method method) noexcept
 {
-  kept_by_block = 0;
+  state = 0;
   do_not_relay = 0;
-  is_local = 0;
+
+  static_assert(unsigned(relay_method::local) == 1, "unexpected relay_method value - see code below");
+  static_assert(unsigned(relay_method::forward) == 2, "unexpected relay_method value - see code below");
 
   switch (method)
   {
     case relay_method::none:
       do_not_relay = 1;
       break;
     case relay_method::local:
-      is_local = 1;
+    case relay_method::forward:
+      state = uint8_t(method) + 1;
+      break;
+    case relay_method::block:
+      state = 1;
       break;
     default:
     case relay_method::fluff:
       break;
-    case relay_method::block:
-      kept_by_block = 1;
-      break;
   }
 }
 
 relay_method txpool_tx_meta_t::get_relay_method() const noexcept
 {
-  if (kept_by_block)
-    return relay_method::block;
+  switch (state)
+  {
+    default:
+    case 0:
+      break;
+    case 1:
+      return relay_method::block;
+    case uint8_t(relay_method::local) + 1:
+      return relay_method::local;
+    case uint8_t(relay_method::forward) + 1:
+      return relay_method::forward;
+   };
```

Comment:
tbh, this is really dodgy. Does this really need to be obfuscated like this ?

Reply:
The code could continue using the inefficient bit mechanism, or could "break" the existing txpool between releases (this code breaks the txpool for users running master branch only).

Reply:
Why do we care about efficient bit twiddling when it's about I/O ?

Reply:
Its less wasted space and less CPU cycles to just store/retrieve this value as the enum value.

Reply:
Well, whatever. If you prefer saving a few cycles hidden next to massive amounts of them for confusing code, fine.

Reply:
No, my preference was to break the existing txpool (pain should be minimal) to do: `m_state = uint8_t(tx_relay);`.

I think in the end, I will drop this. This creates a hard dependency on the enum value and DB value. How about something like:

```c++
const uin8_t state =
  uint8_t(do_not_relay) +
  uint8_t(kept_by_block) << 1 +
  uint8_t(forward) << 2;

switch (state)
{
  default:
  case 0:
    return relay_method::fluff;
  case 1:
    return relay_method::none;
  case 2:
    return relay_method::block;
  case 4:
    return relay_method::forward:
}
```
which also handles the incorrect DB state, and maps it to fluff. Or the if-else ladder (which maps the incorrect DB state to some other value - but I doubt that matters anyway).

Reply:
I much prefer this since I find it much clearer. It looks equivalent to the even simpler:

if (forward)
  return fluff;
if (kept_by_block)
  return block;
if (do_not_relay)
  return none;
return fluff;

Reply:
Amusingly, it looks close to what you had before. The thing I was finding objectionable was not the set of tests, it was the case on enum+1.

Reply:
Yes, the original version was iffy because it required knowledge of the enum values to read it correctly. And a change in enum values could break the function. This version is similar looking, but everything is self-contained within the function.

FWIW, I disassembled this version and gcc 9 was able to optimize the function into a single comparison for the "error" case. Otherwise it used adds, shifts and a 64-bit constant to put the enum value in the return register. Roughly half of the instructions could be done in parallel by the CPU backend.

The if-ladder is less x64-64 instructions but 3-total comparison instructions. The compiler can still generate the same ASM with the switch version, but chose one with more math instructions and less comparisons.

Reply:
For clarity, I really don't care about a dozen cycles saved for this. What I care about is the complexity of the C code.